polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -789,362 +665,335 @@ impl FragmentTree {
-	// av-cores do not back parachain forks), None is returned otherwise.
-	// If we cannot use all ancestors, stop at the first found hole in the chain. This usually
-	// translates to a timed out candidate.
-	fn find_ancestor_path(&self, mut ancestors: Ancestors) -> Option<NodePointer> {
-		// The number of elements in the path we've processed so far.
-		let mut depth = 0;
-		let mut last_node = NodePointer::Root;
-		let mut next_node: Option<NodePointer> = Some(NodePointer::Root);
-
-		while let Some(node) = next_node {
-			if depth > self.scope.max_depth {
-				return None;
+	// Checks if this candidate could be added in the future to this chain.
+	// This assumes that the chain does not already contain this candidate. It may or may not be
+	// present in the `CandidateStorage`.
+	// Even if the candidate is a potential candidate, this function will indicate that it can be
+	// kept only if there's enough room for it.
+	pub(crate) fn can_add_candidate_as_potential(
+		&self,
+		storage: &CandidateStorage,
+		candidate_hash: &CandidateHash,
+		relay_parent: &Hash,
+		parent_head_hash: Hash,
+		output_head_hash: Option<Hash>,
+	) -> PotentialAddition {
+		// If we've got enough candidates for the configured depth, no point in adding more.
+		if self.chain.len() > self.scope.max_depth {
+			return PotentialAddition::None
+		}
+
+		if !self.check_potential(relay_parent, parent_head_hash, output_head_hash) {
+			return PotentialAddition::None
+		}
+
+		let present_in_storage = storage.contains(candidate_hash);
+
+		let unconnected = self
+			.find_unconnected_potential_candidates(
+				storage,
+				present_in_storage.then_some(candidate_hash),
+			)
+			.len();
+
+		if (self.chain.len() + unconnected) < self.scope.max_depth {
+			PotentialAddition::Anyhow
+		} else if (self.chain.len() + unconnected) == self.scope.max_depth {
+			// If we've only one slot left to fill, it must be filled with a connected candidate.
+			PotentialAddition::IfConnected
+		} else {
+			PotentialAddition::None
+		}
+	}
+
+	// The candidates which are present in `CandidateStorage`, are not part of this chain but could
+	// become part of this chain in the future. Capped at the max depth minus the existing chain
+	// length.
+	// If `ignore_candidate`` is supplied and found in storage, it won't be counted.
+	pub(crate) fn find_unconnected_potential_candidates(
+		&self,
+		storage: &CandidateStorage,
+		ignore_candidate: Option<&CandidateHash>,
+	) -> Vec<CandidateHash> {
+		let mut candidates = vec![];
+		for candidate in storage.candidates() {
+			if let Some(ignore_candidate) = ignore_candidate {
+				if ignore_candidate == &candidate.candidate_hash {
+					continue
+				}
+			}
+			// We stop at max_depth + 1 with the search. There's no point in looping further.
+			if (self.chain.len() + candidates.len()) > self.scope.max_depth {
+				break
 			}
+			if !self.candidates.contains(&candidate.candidate_hash) {
+				if self.check_potential(
```

Comment:
Can we merge both conditions in single if ?

Reply:
yes